# Spam Detector

In [49]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [50]:
# Import the data
# I also checked and there are 4601 entries and no NaNs
# All Dtypes are floats or ints
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.info()
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

The [UCI website](https://archive.ics.uci.edu/dataset/94/spambase) seemingly gives away the answer by demonstrating Random Forest's better performance than Logistic Regression (by almost non-overlapping confidence intervals). I therefore have to go with Random Forest as my guess for this exercise.

## Split the Data into Training and Testing Sets

In [51]:
# Create the labels set `y` and features DataFrame `X`
X = data.drop(columns=['spam'])
y = data['spam']
X.info()
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

In [52]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [53]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1173            0.00                0.0           0.00           0.0   
1954            0.67                0.0           0.00           0.0   
256             0.00                0.0           1.42           0.0   
3341            0.00                0.0           0.24           0.0   
4469            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1173           0.00            0.00              1.23                0.00   
1954           0.00            0.00              0.00                0.00   
256            0.71            0.00              0.00                0.71   
3341           0.09            0.04              0.00                0.00   
4469           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1173             0.00            0.00  ...                   0.0        0.000   
1954             0.00            0.00  ...                   0.0        0.000   
256              0.00            0.71  ...                   0.0        0.000   
3341             0.04            0.00  ...                   0.0        0.014   
4469             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1173        0.000          0.0        1.939        0.000        0.000   
1954        0.055          0.0        0.000        0.000        0.000   
256         0.000          0.0        0.931        0.000        0.000   
3341        0.148          0.0        0.014        0.044        0.007   
4469        0.000          0.0        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
1173                       1.379                           8   
1954                       1.644                          13   
256                       12.055                         117   
3341                       2.112                          26   
4469                       1.500                           2   

      capital_run_length_total  
1173                        40  
1954                        74  
256                        217  
3341                      1223  
4469                         3  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [54]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

In [55]:
# Fit the Standard Scaler with the training data
X_train_scaled = scaler.fit_transform(X_train)

# Display the arrays. 
X_train_scaled

array([[-0.33933373, -0.16214839, -0.55144324, ..., -0.11649103,
        -0.21175305, -0.41909144],
       [ 1.8777157 , -0.16214839, -0.55144324, ..., -0.10814021,
        -0.18841316, -0.35960578],
       [-0.33933373, -0.16214839,  2.2325961 , ...,  0.21993678,
         0.29705664, -0.10941609],
       ...,
       [-0.33933373, -0.16214839, -0.55144324, ..., -0.08642808,
        -0.20241709, -0.40334524],
       [-0.33933373,  0.02143809, -0.55144324, ..., -0.10902256,
        -0.18841316, -0.18639753],
       [-0.00843083, -0.16214839,  0.25239911, ..., -0.09960031,
        -0.1137255 ,  0.10403246]])

In [56]:
# Scale the training (*TEST?*) data
X_test_scaled = scaler.transform(X_test)

# Display the arrays. 
X_test_scaled

array([[-0.33933373, -0.16214839, -0.55144324, ..., -0.12843428,
        -0.2444289 , -0.48382584],
       [ 2.01007686, -0.16214839,  0.84057643, ..., -0.12742588,
        -0.23976092, -0.43308807],
       [-0.33933373, -0.16214839,  1.23269465, ..., -0.11835027,
        -0.21642103, -0.30886801],
       ...,
       [-0.1077017 ,  0.01345607,  1.05624145, ..., -0.05189034,
         0.09633355,  1.21501465],
       [-0.33933373, -0.16214839, -0.55144324, ..., -0.10952676,
        -0.23042497, -0.47507794],
       [-0.33933373, -0.16214839, -0.23774866, ..., -0.07788818,
        -0.11839347,  0.83710575]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [57]:
# Create and fit a Logistic Regression model
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(random_state=1)

# fit it to the training data
lr_model = model.fit(X_train_scaled, y_train)

In [58]:
# Make and save testing predictions with the trained Logistic Regression model using the test data
testing_predections_lr = lr_model.predict(X_test_scaled)

# Review the predictions
testing_predections_lr

array([0, 0, 0, ..., 1, 0, 0])

In [59]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predections_lr)

0.9226759339704604

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [60]:
# Create and fit a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(random_state=1)

rf_model = model.fit(X_train_scaled, y_train)

In [61]:
# Make and save testing predictions with the trained Random Forest Classifier model using the test data
testing_predections_rf = rf_model.predict(X_test_scaled)

# Review the predictions
testing_predections_rf

array([0, 0, 0, ..., 1, 0, 0])

In [62]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predections_rf)

0.9582971329278888

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*Logistic Regression: Accuracy looks good, approximately 92% of the emails in the test data were accurately categorized by the model.*

*Random Forest: Accuracy looks extremely good, approximately 96% of the emails in the test data were accurately categorized by the model. I do, though, have some concerns about overfitting given that is a known behavior of Random Forest*